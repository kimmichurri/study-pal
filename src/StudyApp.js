import React, { Component } from 'react';
import brainLogo from './assets/brainimg.png';
import Header from './Header'
import Flashcard from './Flashcard'
import Greeting from './Greeting'
import { stringAndArrayPrototypeMethods } from './PrototypeMethodsData'


export default class StudyApp extends Component {
    constructor() {
        super();
        this.state = {
            flashcardData: stringAndArrayPrototypeMethods,
            // in component did mount set the value of flashcardData to the whole dataset
            flashcards: [],
            studyBank: JSON.parse(localStorage.getItem("study bank cards")) || [],
            currentFlashcardIndex: 0
            // if study bank is empty disable the button, if study bank has length we will pull flashcards from local storage when the button is clicked
        }
    }

    chooseStudyBank = (e) => {
        console.log('click', e.target)
        console.log(this.state.flashcards)
        const studyBanksCards = [...this.state.studyBank]
        if (e.target.innerText === 'Study Bank') {
            this.setState({
                flashcards: studyBanksCards,
                currentFlashcardIndex: 0
            })
        }
    }

    chooseArrayCategory = (e) => {
        if (e.target.innerText === 'Array Prototype Methods') {
            let arrayQuestions = this.state.flashcardData.filter((question) => {
                return question.category === "Array Methods"
            })
            this.setState({
                flashcards: arrayQuestions
            })
        }   
    }

    chooseStringCategory = (e) => {
        if (e.target.innerText === 'String Prototype Methods') {
            let stringQuestions = this.state.flashcardData.filter((question) => {
                return question.category === "String Methods"
            })
            this.setState({
                flashcards: stringQuestions
            })
        }   
    }

    incrementCurrentIndex = () => {
        this.setState({
            currentFlashcardIndex: this.state.currentFlashcardIndex + 1
        })
    }

    render() {
        let numberOfStudyQuestions = this.state.studyBank.length;
        return (
            <div>
                <Header />
                <img src={brainLogo} className="study-pal-logo" alt="brain-logo" />
                <section className= "study-choice-buttons-container">
                    <button onClick={this.chooseStudyBank} className="study-choice-buttons">Study Bank ({numberOfStudyQuestions})</button>
                    <button onClick={this.chooseArrayCategory} className="study-choice-buttons">Array Prototype Methods</button>
                    <button onClick={this.chooseStringCategory} className="study-choice-buttons">String Prototype Methods</button>
                </section>
                {this.state.flashcards.length &&
                <Flashcard 
                    flashcards={this.state.flashcards}
                    studyBank={this.state.studyBank}
                    currentIndex={this.state.currentFlashcardIndex}
                    incrementCurrentIndex={this.incrementCurrentIndex}
                />
                }
                {!this.state.flashcards.length &&
                <Greeting />
                }
            </div>
        )
    }
}