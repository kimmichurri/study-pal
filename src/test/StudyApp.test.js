import React from 'react';
import StudyApp from '../StudyApp';
import { shallow } from 'enzyme';

describe('StudyApp', () => {

    let wrapper;

    const chooseArrayCategoryMock = jest.fn();

    beforeEach(() => {

        wrapper = shallow(
            <StudyApp 
            chooseArrayCategory={chooseArrayCategoryMock}
            />
        )
    });

    it('should have the proper default state', () => {
        expect(wrapper.state()).toEqual({
            flashcardData: [],
            flashcards: [],
            studyBank: [],
            currentFlashcardIndex: 0
        })
    });

    it('should reassign flashcards to study bank when chooseStudyBank function is invoked', () => {
        const mockStudyBank = [
            {
                "id": "a257d840-36c6-11e9-b210-d663bd873d93",
                "category": "String Methods",
                "question": "In the following code what does 'gi' indicate?: var newstr = str.replace(/apples/gi, 'oranges')",
                "answerChoices": ["Replace all 'g' and 'i' characters in the string", "Replace 'apples' with 'oranges'", "Replace all 'apples' with 'oranges' and ignore letter case"],
                "correctAnswer": "Replace all 'apples' with 'oranges'",
                "link": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace"
            }
        ]
        expect(wrapper.state('flashcards')).toEqual([]);
        wrapper.instance().state.studyBank = mockStudyBank;
        wrapper.instance().chooseStudyBank();
        expect(wrapper.state('flashcards')).toEqual(mockStudyBank);
    });

    it('should filter an array of string method flashcards when chooseStringCategory is invoked', () => {
        const mockFlashcards = [
            {
                "id": "a257df02-36c6-11e9-b210-d663bd873d93",
                "category": "String Methods",
                "question": "I am given the string 'Turing School of Software and Design' and I need to return an array with each word as an element. Which syntax of .split() will accomplish this?",
                "answerChoices": [".split(' ')", ".split('')", ".split()"],
                "correctAnswer": ".split(' ')",
                "link": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split"
            },
            {
                "id": "a257e1aa-36c6-11e9-b210-d663bd873d93",
                "category": "Array Methods",
                "question": "I typed in 'Kim Myers'.join() to join these two words into one and received an error message. Why is that?",
                "answerChoices": [".join() is an array method and cannot be used on a string", ".join() is missing quotation marks within the parenthesis", "You need to store the string in a variable before using .join()"],
                "correctAnswer": ".join() is an array method and cannot be used on a string",
                "link": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join"
            }
        ]
        expect(wrapper.state('flashcards')).toEqual([]);
        wrapper.instance().state.flashcardData = mockFlashcards;
        wrapper.instance().chooseStringCategory();
        expect(wrapper.state('flashcards')).toEqual([mockFlashcards[0]]);
    });

    it('should increment flashcard index by 1 when incrementCurrentIndex is invoked', () => {
        expect(wrapper.state('currentFlashcardIndex')).toEqual(0);
        wrapper.instance().incrementCurrentIndex();
        expect(wrapper.state('currentFlashcardIndex')).toEqual(1);
    });

    it('should display header and greeting by default', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it.skip('should display flashcard if the user has selected flashcards', () => {
        const mockFlashcards = [
            {
                "id": "a257e1aa-36c6-11e9-b210-d663bd873d93",
                "category": "Array Methods",
                "question": "I typed in 'Kim Myers'.join() to join these two words into one and received an error message. Why is that?",
                "answerChoices": [".join() is an array method and cannot be used on a string", ".join() is missing quotation marks within the parenthesis", "You need to store the string in a variable before using .join()"],
                "correctAnswer": ".join() is an array method and cannot be used on a string",
                "link": "https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/join"
            }
        ]
        wrapper.find('.array-button').simulate('click');
        expect(chooseArrayCategoryMock).toBeCalled();
        wrapper.instance().state.flashcards = mockFlashcards;
        expect(wrapper).toMatchSnapshot();
    })



});