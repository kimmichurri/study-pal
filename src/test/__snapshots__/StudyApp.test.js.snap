// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StudyApp should display flashcard if the user has selected flashcards 1`] = `
<div>
  <Header />
  <img
    alt="brain-logo"
    className="study-pal-logo"
    src="brainimg.png"
  />
  <section
    className="study-choice-buttons-container"
  >
    <button
      className="study-choice-buttons array-button"
      onClick={[Function]}
    >
      Array Prototype Methods
    </button>
    <button
      className="study-choice-buttons string-button"
      onClick={[Function]}
    >
      String Prototype Methods
    </button>
    <button
      className="study-choice-buttons study-bank-button"
      onClick={[Function]}
    >
      Study Bank (
      0
      )
    </button>
  </section>
  <Greeting />
</div>
`;

exports[`StudyApp should display header and greeting by default 1`] = `
<div>
  <Header />
  <img
    alt="brain-logo"
    className="study-pal-logo"
    src="brainimg.png"
  />
  <section
    className="study-choice-buttons-container"
  >
    <button
      className="study-choice-buttons array-button"
      onClick={[Function]}
    >
      Array Prototype Methods
    </button>
    <button
      className="study-choice-buttons string-button"
      onClick={[Function]}
    >
      String Prototype Methods
    </button>
    <button
      className="study-choice-buttons study-bank-button"
      onClick={[Function]}
    >
      Study Bank (
      0
      )
    </button>
  </section>
  <Greeting />
</div>
`;
